cmd/docker/docker.go

```diff
@@ -228,6 +231,34 @@ func initializeDockerCli(dockerCli *command.DockerCli, flags *pflag.FlagSet, opt
 	return dockerCli.Initialize(opts)
 }
 
+func executeCmd(dockerCli *command.DockerCli, cmd *cobra.Command) {
+	if err := cmd.Execute(); err != nil {
+		switch err := err.(type) {
+		case cli.StatusError:
+			{
+				if err.Status != "" {
+					fmt.Fprintln(dockerCli.Err(), err.Status)
+				}
+				// StatusError should only be used for errors, and all errors should
+				// have a non-zero exit status, so never exit with 0
+				if err.StatusCode == 0 {
+					os.Exit(1)
+				}
+				os.Exit(err.StatusCode)
+			}
+		case *exec.ExitError:
+			{
+				if ws, ok := err.Sys().(syscall.WaitStatus); ok {
+					os.Exit(ws.ExitStatus())
+				}
+				os.Exit(1)
+			}
```

Comment:
I think it would be cleaner to have the code in `newDockerCommand.RunE` which calls `plugincmd.Run()` and returns the result to capture the `exec.ExitError` and convert it into a `cli.StatusError`. That way the knowledge that some commands are external does not leak any further that it already has.

Reply:
Agreed